### DISCUSSION TITLE
Calculations

### DISCUSSION THREAD

**Discussion by alanruttenberg** [CommonCoreOntology/CommonCoreOntologies#385] [Ontology Terms and Class Schema]
I don't see a term equivalent to OBI's [data transformation](http://purl.obolibrary.org/obo/OBI_0200000).

If there is, please respond with the label.

If not, I'd like to request a similar term be added to CCO.

I'll need this as I start adding terms about image processing in the ontology I am working on.

Connecting it to CCO: A number of the artifact functions or subclasses of them are realized in a data transformations, e.g. **Signal Processing Artifact Function**. I think. The uncertainty is because there is, e.g., optical signal processing that isn't at the information level. I don't know if such processes are covered by the definition. If not, it wouldn't be a subclass of data transformation. Processes involving **Information Processing Artifacts** would take **Information Content Entities** as **input** and produce them as **output**.

If we are at all considering incorporating OBO terms, data transformations and some related terms would be good candidates.

Tagging @karenyork, who is working with me.

**Comment by nklsbckmnn** [comment]
Does "algorithm" fit the bill?

Edit: One could have a "computation" process inputs and outputs of which map to inputs and outputs prescribed by an algorithm.

There is also "computing artifact function" that refers to a computation process.

**Comment by alanruttenberg** [comment]
The request was for the process itself. I know I can define a process using the functions available and algorithm. I think it's a common enough process that it belongs in the ontology. Computing Artifact Function refers to a process only in its English definition, but there is no axiom that relates them.  Have a look at OBI's [data transformation](http://purl.obolibrary.org/obo/OBI_0200000) branch.

Both the functions and the algorithms are different components of definition. Sometimes you know the algorithm sometimes not. Sometimes a computation realizes one function, sometimes more than one. The same algorithm may realize different functions. For instance a rescaling computation may be used to condition data or to normalize it. OBI also uses [achieves_planned_objective](http://purl.obolibrary.org/obo/OBI_0000417) and [data transformation objective](http://purl.obolibrary.org/obo/OBI_0200166) in defining computation processes. 

Having the process also allows the input and output types to be specified. It isn't entirely clear whether Computing Artifact Function always yields data as output. For instance it would seem to me that a spln loop used to create a delay, for instance to rate limit password attempts would be a realization of a Computing Artifact Function.  The data transformation process as OBI has defined it is unambiguous on that question.

Finally, experience shows that without creating compositional definitions of computing processes there will be asserted polyhierarchy. Just defining the data transformation process doesn't fix that but it can help establish a pattern for how to represent computations. 

Thinking ahead to a place where we can interoperate well with OBO, I'd like to consider aligning this part of CCO with OBO's as OBI defines a bunch of computation terms that can be reused instead of defining the terms de-novo. In order to do that we would need to work on aligning CCO's IAO with IAO and OBI's - there's a lot of duplication, and consider adopting OBO's [has_specified_input](http://purl.obolibrary.org/obo/OBI_0000293) and [has_specified_output](http://purl.obolibrary.org/obo/OBI_0000299), which I consider more clearly defined than the current has_input and has_output relations.

**Comment by cameronmore** [comment]
@alanruttenberg does #211 capture the intended process here?

**Comment by neilotte** [comment]
@alanruttenberg @cameronmore  Converting to a discussion thread. See also a gloss here: https://github.com/CommonCoreOntology/CommonCoreOntologies/discussions/384
